### ISSUE TITLE
Referencing the result of an array append operation

### ISSUE THREAD

**Issue by derrabus** [vimeo/psalm#5287]
PHP allows to assign the result of an array append operation by reference, for example:

```php
/**
 * @param list<int> $collection
 * @return list<int>
 */
function add42(array $collection): array {
    $item = &$collection[];
    $item = 42;
    
    return $collection;
}
```

https://psalm.dev/r/b430746f42
https://3v4l.org/1L21Q

This might not be the most elegant way to append a value to an array, yet it works. Psalm complains:

```
INFO: UnusedVariable - 8:5 - $item is never referenced or the value is not used
```

I would consider this to be a false positive.

**Comment by derrabus** [comment]
Taking this a step further:

```php
/**
 * @param array{numbers?: list<int>} $collection
 * @return array{numbers: list<int>}
 */
function add42(array $collection): array {
    $item = &$collection['numbers'][];
    $item = 42;
    
    return $collection;
}
```

https://psalm.dev/r/9158ba8802
https://3v4l.org/SWklB

Here, Psalm complains about a `PossiblyUndefinedArrayOffset`. This is not correct imo: The append operation will create the array offset `numbers` if it does not exist already.

All those errors go away if I directly assign the value to the append operator:

https://psalm.dev/r/0197a64131

**Comment by muglug** [comment]
This is both a bug, and a hard `wontfix` from me.

I've never seen code that does this, and looking at the code doesn't really tell me what it _should_ do.

I think that if I have to run the code to figure out how it behaves, that is bad code that Psalm should flag.

If you love this pattern and want Psalm to understand it, feel free to create a PR.

**Comment by derrabus** [comment]
> looking at the code doesn't really tell me what it _should_ do.

I 100% agree! I myself was also pretty puzzled when I discovered it.

> I've never seen code that does this

Now you have:

https://github.com/symfony/symfony/blob/e2f1c467348952ab9dbe31b473dcf35015527f1b/src/Symfony/Component/EventDispatcher/EventDispatcher.php#L263-L274

The code in question needs to keep a reference to the element that is about to be added to the array for other purposes. In this particular case, we add a lazy proxy to the array that replaces itself when being accessed for the first time.

> If you love this pattern

I absolutely don't! 😅 

> and want Psalm to understand it, feel free to create a PR.

Since I haven't contributed to Psalm's codebase yet, I would need some pointers where to start.

**Comment by muglug** [comment]
@derrabus is there a reason the Symfony example wouldn't work with 

```php
$this->optimized[$eventName][] = $closure;
```

after the `if/else`?

**Comment by derrabus** [comment]
Yes, the lazy proxy would not work:

https://3v4l.org/loKII Vs. https://3v4l.org/U6TOh

**Comment by weirdan** [comment]
@derrabus This (https://3v4l.org/bP964) would work though.

**Comment by derrabus** [comment]
@weirdan Yes, that would work.

Okay, so we agree that the code in question would not win a beauty contest. The question remains: do you see value in teaching Psalm to understand it? If so, I'd like to try if I can contribute that feature, but I might need some help.

**Comment by muglug** [comment]
> The question remains: do you see value in teaching Psalm to understand it? If so, I'd like to try if I can contribute that feature, but I might need some help.

I don't see value to anyone beyond you/the Symfony project, as I'm not aware of that pattern being used anywhere else. And if Symfony's going to be using the baseline, you can pretty much just ignore this in perpetuity.

Generally my approach with these sorts of edge-cases is: "can this be rewritten in a way that's a little clearer to humans and much more clear to Psalm? If so, do that instead of trying to make Psalm understand the pattern".

**Comment by muglug** [comment]
As to the implementation, it's tricky. References are hard to analyse statically ([here's a sample paper on the topic](https://www.cs.cornell.edu/courses/cs711/2005fa/papers/steensgaard-popl96.pdf)), especially in interpreted languages. That's why Facebook basically scrapped them altogether when they forked PHP into Hack.

**Comment by derrabus** [comment]
All right, thank you for your assessment. I won't pursue this issue for now then. 🙂
